materialize-snowflake/snowflake.go

```diff
@@ -610,79 +632,158 @@ func (d *transactor) Store(it *pm.StoreIterator) (pm.StartCommitFunc, error) {
 		} else if err = b.store.stage.encodeRow(converted); err != nil {
 			return nil, fmt.Errorf("encoding Store to scratch file: %w", err)
 		}
+	}
 
-		if it.Exists {
-			b.store.mustMerge = true
+	// Upload the staged files and build a list of merge and copy into queries that need to be run
+	// to effectively commit the files into destination tables. These queries are stored
+	// in the checkpoint so that if the connector is restarted in middle of a commit
+	// it can run the same queries on the next startup. This is the pattern for
+	// recovery log being authoritative and the connector idempotently applies a commit
+	// These are keyed on the binding table name so that in case of a recovery being necessary
+	// we don't run queries belonging to bindings that have been removed
+	var queries = make(map[string][]string)
+	// all files uploaded across bindings
+	var toDelete = make(map[string][]string)
+
+	log.Info("store: starting copying of files into tables")
+	for _, b := range d.bindings {
+		if !b.store.stage.started {
+			continue
 		}
+
+		dir, err := b.store.stage.flush()
+		if err != nil {
+			return nil, err
+		}
+
+		if v, ok := queries[b.target.Identifier]; ok {
+			if b.target.DeltaUpdates {
+				queries[b.target.Identifier] = append(v, renderWithDir(b.store.copyInto, dir))
+			} else {
+				queries[b.target.Identifier] = append(v, renderWithDir(b.store.mergeInto, dir))
```

Comment:
This scenario:

We run a transaction up to the point of rendering the queries, committing to the runtime recovery log and before we can actually apply the query "fully", the connector is restarted. That is, of the N queries we have to run for this transaction, we ran half of them, but missed the other half. If there is a `COPY INTO` query among the half that was already run, on restart of the connector, we will end up running another `COPY INTO` which will lead to duplicate data

Reply:
Are the `COPY INTO` queries idempotent? I had thought they wouldn't copy the same files twice. Otherwise, could that scenario happen for a delta updates materialization too?

Reply:
@williamhbaker ah yeah, this might actually work. I was testing switching back and forth between connector versions and got confused about idempotency of `COPY INTO`, but in the case of the same files being used, `COPY INTO` is supposed to be idempotent. I'll try this out

Reply:
Tested it out, it is indeed idempotent. Had an artificial error cause the connector to restart after applying checkpoint multiple times, with the same `COPY INTO` query, and only one of them was applied.